relayer/chain/ethereum/header_chain_state.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package ethereum
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/tranvictor/ethashproof"
+	"golang.org/x/sync/errgroup"
+)
+
+type EthashproofCacheState struct {
+	sync.Mutex
+	currentCache *ethashproof.DatasetMerkleTreeCache
+	nextCache    *ethashproof.DatasetMerkleTreeCache
+}
+
+// HeaderChainState is a helper for tracking where we are
+// in the Ethereum chain. It warms up caches as we go.
+type HeaderChainState struct {
+	currentBlock          uint64
+	newestBlock           uint64
+	ethashproofCacheState *EthashproofCacheState
+	eg                    *errgroup.Group
+}
+
+func NewHeaderChainState(eg *errgroup.Group, currentBlock uint64, newestBlock uint64) (*HeaderChainState, error) {
+	cacheState := EthashproofCacheState{
+		currentCache: nil,
+		nextCache:    nil,
+	}
+	state := HeaderChainState{
+		currentBlock:          currentBlock,
+		newestBlock:           newestBlock,
+		ethashproofCacheState: &cacheState,
+		eg:                    eg,
+	}
+
+	// Block until cache for current epoch is prepared
+	cache, err := makeEthashproofCache(currentBlock / 30000)
+	if err != nil {
+		return nil, err
+	}
+	cacheState.currentCache = cache
+	// Asynchronously prepare next epoch's cache
+	eg.Go(func() error {
+		return state.prepareNextEthashproofCache()
+	})
+
+	return &state, nil
+}
+
+// GetEthashProofCache returns the cache used for proof generation. It will return
+// immediately if `number` is in the current or next epoch. Outside that range, it
+// might block for multiple minutes to generate the cache. Calling GetEthashproofCache
+// will also update the current epoch to `number` / 30000.
+func (s *HeaderChainState) GetEthashproofCache(number uint64) (*ethashproof.DatasetMerkleTreeCache, error) {
```

Comment:
This logic should really have a test. @vgeddes @musnit is there a mocking library you folks use?

Reply:
Agree, tests would be great for this file. 

I prefer https://github.com/stretchr/testify over https://github.com/golang/mock for the better output, but both are fine in my view.